File: 04/ruby/part01.rb
# Day 4: Secure Container
# https://adventofcode.com/2019/day/4#part2

def sorted?(chars)
  chars.sort == chars
end

def paired?(chars)
  lookup = Hash.new { |h, k| h[k] = 0 }
  chars.each { |char| lookup[char] += 1 }
  lookup.values.any? { |count| count > 1 }
end

a, b = File.read('./inputs/04.txt').split('-').map(&:to_i)
count = 0

while a <= b
  bytes = a.to_s.bytes
  count += 1 if sorted?(bytes) && paired?(bytes)
  a += 1
end

puts count

File: 04/nim/part01.nim
# Day 4: Secure Container
# https://adventofcode.com/2019/day/4

import algorithm, strutils, sequtils, tables

proc paired(nums: seq[char]): bool =
  let numCount = toCountTable(nums)
  result = false
  for _, count in numCount.pairs:
    if count > 1:
      result = true
      return

let
  input = open("./inputs/04.txt")
  vals  = input.readAll().split({ '-' }).mapIt(parseInt(strip(it)))
  a     = vals[0]
  b     = vals[1]

input.close()

var count = 0

for num in a..b:
  let parts = toSeq($num)
  if isSorted(parts) and paired(parts):
    count += 1

echo count
